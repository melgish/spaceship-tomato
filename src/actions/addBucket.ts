'use server'
import { revalidatePath } from 'next/cache'
import { prisma } from '@/lib/prisma'
import { type AddBucket, addBucketSchema } from '@/schemas/addBucket.schema'

/**
 * Add server action to a new bucket to the database.
 * @see ../schemas/addBucket.schema.ts for validation rules.
 * @param payload The request payload
 * @returns Ok result when successful, otherwise an Err result.
 */
export async function addBucket(payload: AddBucket) {
  try {
    const data = await addBucketSchema.parseAsync(payload)
    const created = await prisma.bucket.create({
      data: { ...data, plantId: data.plantId || null },
    })
    revalidatePath('/buckets')
    return { ok: true, message: 'Bucket created', data: created } as const
  } catch (error: any) {
    return { ok: false, message: 'Failed to create bucket' } as const
  }
}
