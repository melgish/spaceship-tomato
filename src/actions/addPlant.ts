'use server'
import { revalidatePath } from 'next/cache'
import { prisma } from '@/lib/prisma'
import { type AddPlant, addPlantSchema } from '@/schemas/addPlant.schema'

/**
 * Add server action to a new plant to the database.
 * @see ../schemas/addPlant.schema.ts for validation rules.
 * @param payload The request payload
 * @returns Ok result when successful, otherwise an Err result.
 */
export async function addPlant(payload: AddPlant) {
  try {
    const data = await addPlantSchema.parseAsync(payload)
    const created = await prisma.plant.create({ data })
    revalidatePath('/plants')
    return { ok: true, message: 'Plant created', data: created } as const
  } catch (error: any) {
    return { ok: false, message: 'Failed to create plant' } as const
  }
}
