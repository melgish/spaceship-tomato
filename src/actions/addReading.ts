'use server'
import { revalidatePath } from 'next/cache'
import { prisma } from '@/lib/prisma'
import { type AddReading, addReadingSchema } from '@/schemas/addReading.schema'

/**
 * Add server action to a new plant to the database.
 * @see ../schemas/addReading.schema.ts for validation rules.
 * @param payload The request payload
 * @returns Ok result when successful, otherwise an Err result.
 */
export async function addReading(payload: AddReading) {
  try {
    const data = await addReadingSchema.parseAsync(payload)
    const created = await prisma.reading.create({ data })
    revalidatePath('/readings')
    return { ok: true, message: 'Reading created', data: created } as const
  } catch (error: any) {
    return { ok: false, message: 'Failed to create reading' } as const
  }
}
