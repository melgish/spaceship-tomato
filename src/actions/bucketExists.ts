'use server'
import { prisma } from '@/lib/prisma'

/**
 * A server action to validate that a bucket exists in the database.
 * @param id Id of bucket to search for.
 * @returns true if bucket exists, otherwise false.
 */
export async function bucketExists(id: number = 0) {
  return id > 0 && (await prisma.bucket.count({ where: { id } })) > 0
}
