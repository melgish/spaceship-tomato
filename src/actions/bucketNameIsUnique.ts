'use server'
import { prisma } from '@/lib/prisma'

/**
 * A server action to validate that the name of a bucket is unique.
 * @param name Name of the bucket to test.
 * @param id Optional id of bucket to exclude when editing an existing bucket.
 * @returns true if the name exists on another bucket, otherwise false.
 */
export async function bucketNameIsUnique(name: string, id: number = 0) {
  const where = { name, id: (id > 0 && { not: id }) || undefined }
  return (await prisma.bucket.count({ where })) === 0
}
