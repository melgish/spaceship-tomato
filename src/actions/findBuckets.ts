'use server'
import { prisma } from '@/lib/prisma'

/**
 * A server action to fetch all of buckets from the database.
 * @returns An array of buckets. The returned data will include plant info if
 * the bucket contains a plant.
 */
export async function findBuckets() {
  return await prisma.bucket.findMany({
    include: { plant: true },
    orderBy: [{ name: 'asc' }],
  })
}

export type Buckets = Awaited<ReturnType<typeof findBuckets>>
export type Bucket = Buckets[number]
