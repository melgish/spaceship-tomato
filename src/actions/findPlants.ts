'use server'
import { prisma } from '@/lib/prisma'

/**
 * A server action to fetch all of plants from the database.
 * @returns An array of plants.
 */
export async function findPlants() {
  return await prisma.plant.findMany({
    orderBy: [{ name: 'asc' }],
  })
}

export type Plants = Awaited<ReturnType<typeof findPlants>>
export type Plant = Plants[number]
