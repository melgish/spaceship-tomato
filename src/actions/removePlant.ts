'use server'
import { prisma } from '@/lib/prisma'
import { revalidatePath } from 'next/cache'

/**
 * A server action to remove a single plant from the system.
 * If the plant is associated with a bucket, that plantId value of that bucket
 * will be set to NULL.
 * @param id Id of the plant to remove.
 * @returns an Ok result if the identified plant was removed, otherwise an
 *  Err result.
 */
export async function removePlant(id: number) {
  try {
    await prisma.plant.delete({ where: { id }, select: { id: true } })
    revalidatePath('/plants')
    return { ok: true, message: 'Plant deleted', data: id } as const
  } catch (error: any) {
    return { ok: false, message: 'Failed to remove plant' } as const
  }
}
