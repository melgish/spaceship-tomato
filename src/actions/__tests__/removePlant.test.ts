import { describe, expect, it } from 'vitest'
import { removePlant } from '@/actions/removePlant'
import { prisma } from '@/lib/__mocks__/prisma'

describe('when bucket is not in database', () => {
  it('should return a result', async () => {
    prisma.plant.delete.mockRejectedValue('boom')

    const rs = await removePlant(1)

    expect(rs).toMatchSnapshot()
  })
})

describe('when plant is in the database', () => {
  it('should return a result', async () => {
    prisma.plant.delete.mockResolvedValue({ id: 1 } as any)

    const rs = await removePlant(1)

    expect(rs).toMatchSnapshot()
  })
})
