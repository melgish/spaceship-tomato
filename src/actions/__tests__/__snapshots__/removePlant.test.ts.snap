// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`when bucket is not in database > should return a result 1`] = `
{
  "message": "Failed to remove plant",
  "ok": false,
}
`;

exports[`when plant is in the database > should return a result 1`] = `
{
  "data": 1,
  "message": "Plant deleted",
  "ok": true,
}
`;
